### PR TITLE
Fix Bouncy Castle certificate validation and add FTPS regression coverage

### PR DESCRIPTION
**Pull request by Wixely**
When I added the BouncyCastle lib for FTPS I didn't have any need for hostnames and never encounted the issues raised in https://github.com/robinrodricks/FluentFTP/issues/1866
There was a real issue where a valid cert could be used for a domain different from the one you were accessing and it would blindly accept it.

The fix for this is to check the hostname/ip against the cert.

I also wanted to make sure all these edge cases worked together, like X1C Printer (using session re-use) and hostname validation, so there are a lot of tests built out to prove this.
So some minor changes to the cypher list we negotiate was made to add compatability without exposing old insecure cyphers.

### CODE REVIEW

File: .github/workflows/bouncycastle.yml [hunk 1]
```diff
@@ -0,0 +1,79 @@
+name: bouncycastle
```
Comment (robinrodricks) [inline .github/workflows/bouncycastle.yml:1]: what is this workflow and why is it required?

File: FluentFTP.Dockers/proftpd/Dockerfile [hunk 2]
```diff
@@ -44,6 +44,7 @@ RUN	sed -i -e "s/\r//" /etc/proftpd/proftpd.conf && \
 	openssl req -x509 -newkey rsa:4096 \
 		-keyout /etc/ssl/private/proftpd.key -out /etc/ssl/certs/proftpd.crt \
 		-subj "/C=US/ST=State/L=/O=Dev/CN=fluentftp" \
+		-addext "subjectAltName=DNS:localhost,DNS:fluentftp" \
```
Comment (robinrodricks) [inline FluentFTP.Dockers/proftpd/Dockerfile:47]: why?

File: FluentFTP.Dockers/pureftpd/Dockerfile [hunk 3]
```diff
@@ -36,6 +36,7 @@ RUN	sed -i -e "s/\r//" /usr/sbin/run-pureftpd.sh && \
 	openssl req -x509 -newkey rsa:4096 \
 		-keyout /etc/ssl/private/pure-ftpd.key -out /etc/ssl/certs/pure-ftpd.crt \
 		-subj "/C=US/ST=State/L=/O=Dev/CN=fluentftp" \
```
Comment (robinrodricks) [inline FluentFTP.Dockers/pureftpd/Dockerfile:38]: why?

File: FluentFTP.Dockers/vsftpd/Dockerfile [hunk 4]
```diff
@@ -44,6 +44,7 @@ RUN	sed -i -e "s/\r//" /etc/vsftpd.conf && \
 	openssl req -x509 -newkey rsa:4096 \
 		-keyout /etc/ssl/private/vsftpd.key -out /etc/ssl/certs/vsftpd.crt \
 		-subj "/C=US/ST=State/L=/O=Dev/CN=fluentftp" \
+		-addext "subjectAltName=DNS:localhost,DNS:fluentftp" \
```
Comment (robinrodricks) [inline FluentFTP.Dockers/vsftpd/Dockerfile:47]: why?

File: FluentFTP.Tests/Integration/System/IntegrationTestRunner.cs [hunk 5]
```diff
@@ -19,6 +19,8 @@ public static async Task Run(FtpServer serverType, UseStream useStream, bool use
 			// spin up a new docker
 			using var server = new DockerFtpServer(serverType, useStream.ToString(), useSsl);
 
+			IntegrationTestSuite.ResetCertificateCounters();
```
Comment (robinrodricks) [inline FluentFTP.Tests/Integration/System/IntegrationTestRunner.cs:22]: why?

File: README.md [hunk 6]
```diff
@@ -85,7 +85,7 @@ Stable binaries are released on NuGet, and contain everything you need to use FT
...
 | **[FluentFTP.GnuTLS](https://www.nuget.org/packages/FluentFTP.GnuTLS)**      	|    [![Version](https://img.shields.io/nuget/vpre/FluentFTP.GnuTLS.svg)](https://www.nuget.org/packages/FluentFTP.GnuTLS)  		| [![Downloads](https://img.shields.io/nuget/dt/FluentFTP.GnuTLS.svg)](https://www.nuget.org/packages/FluentFTP.GnuTLS) | [GnuTLS Docs](https://github.com/robinrodricks/FluentFTP/wiki/FTPS-Connection-using-GnuTLS) |
-| **FluentFTP.BouncyCastle** | Fully managed TLS session reuse | Bouncy Castle TLS | [Bouncy Castle Docs](FluentFTP.BouncyCastle/README.md) |
+| **[FluentFTP.BouncyCastle](https://www.nuget.org/packages/FluentFTP.BouncyCastle)**       |    [![Version](https://img.shields.io/nuget/vpre/FluentFTP.BouncyCastle.svg)](https://www.nuget.org/packages/FluentFTP.BouncyCastle)    | [![Downloads](https://img.shields.io/nuget/dt/FluentFTP.BouncyCastle.svg)](https://www.nuget.org/packages/FluentFTP.BouncyCastle) | [Bouncy Castle Docs](FluentFTP.BouncyCastle/README.md) |
```
Comment (robinrodricks) [inline README.md:88]: ok

**Comment by robinrodricks** [comment]
Still awaiting an update.

**Comment by Wixely** [comment]
Hi @robinrodricks, I've removed a lot of the code that encroached on the original tests, this was useful for local testing but I appreciate that they don't make sense committing. Please have a look now and if you want me to handle any of the outdated comments I will do that.

**Comment by robinrodricks** [comment]
Merging since you're the original author and you are performing your own testing.

There are so many changes in this PR that its humanly impossible to review. I will just have to trust you.

**Comment by Wixely** [comment]
Thanks @robinrodricks and sorry for the inconvenience. About 50% of it is tests, I wanted to be very thorough on it this time because I felt like I had overlooked something basic on the first PR.

**Comment by robinrodricks** [comment]
No problem and glad you followed thru. When do you want this released to nuget?

**Comment by Wixely** [comment]
@robinrodricks I did some more testing on the final merged build and everything seemed good. I'm happy for it to get released when you are. In your own time really. When you do that I will update my [mcp server](https://github.com/Wixely/BambuMCPSharp) to use the nuget instead of my adapter [library](https://github.com/Wixely/Wixely.FluentFTP.BouncyCastle).

**Comment by robinrodricks** [comment]
Docs in markdown is not allowed with this repo. I have moved ur docs here:

https://github.com/robinrodricks/FluentFTP/wiki/FTPS-Connection-using-BouncyCastle

**Comment by robinrodricks** [comment]
Released. Thanks for this major feature.

https://www.nuget.org/packages/FluentFTP/55.0.0
https://www.nuget.org/packages/FluentFTP.BouncyCastle/1.0.0
